File: src/model_eval.cpp
#include <absl/flags/flag.h>
#include <absl/flags/parse.h>
#include <absl/log/globals.h>
#include <absl/log/initialize.h>
#include <absl/log/log.h>
#include <absl/log/log_sink_registry.h>

#include "alpha_zero.h"

// new flag specific to this binary

// existing flags from main.cpp
ABSL_FLAG(int, channel_size, 128, "Channel size (must be power of 2)");
ABSL_FLAG(int, num_simulations, 200, "Number of MCTS simulations per move");
ABSL_FLAG(int, batch_size, 1000, "Number of nodes to process at once");
ABSL_FLAG(int, eval_timeout_ms, 1,
          "Timeout for evaluation requests in milliseconds");
ABSL_FLAG(int, num_games, 100, "Number of games to play in evaluation.");
ABSL_FLAG(std::string, model_path, "model.pt", "Path to the model file");
ABSL_FLAG(std::string, training_file, "training_data.bin",
          "File to store training data");
ABSL_FLAG(bool, debug, false,
          "Enable debug logging and intermediate state dumps");

namespace alphazero {
void init_globals() {
  g_config.channel_size = absl::GetFlag(FLAGS_channel_size);
  g_config.num_simulations = absl::GetFlag(FLAGS_num_simulations);
  g_config.batch_size = absl::GetFlag(FLAGS_batch_size);
  g_config.eval_timeout =
      duration<int, milli>(absl::GetFlag(FLAGS_eval_timeout_ms));
  g_config.num_games = absl::GetFlag(FLAGS_num_games);
  g_config.model_path = absl::GetFlag(FLAGS_model_path);
  g_config.training_file = absl::GetFlag(FLAGS_training_file);
  g_config.debug = absl::GetFlag(FLAGS_debug);
  // initialize the evaluation queue with the channel size
  g_evaluation_queue = std::make_unique<eval_channel_t>(g_config.channel_size);
}

}  // namespace alphazero

int main(int argc, char* argv[]) {
  absl::ParseCommandLine(argc, argv);
  absl::InitializeLog();
  absl::SetMinLogLevel(absl::LogSeverityAtLeast::kInfo);
  absl::SetStderrThreshold(absl::LogSeverityAtLeast::kInfo);
  alphazero::FileSink file_sink("model_eval.log");
  absl::AddLogSink(&file_sink);

  alphazero::init_globals();

  LOG(INFO) << "Starting AlphaZero agent tournament...";

  // Start evaluator thread
  std::thread evaluator_thread(&alphazero::run_evaluator);

  // Run the agent tournament
  alphazero::run_agent_tournament();

  // Signal evaluator to stop and join thread
  alphazero::g_stop_evaluator = true;
  evaluator_thread.join();

  return 0;
}

File: scripts/model_eval.py
#!/usr/bin/env python3
import argparse
import min_alpha_zero as maz


def main():
    parser = argparse.ArgumentParser(description="Model evaluation against agents")
    parser.add_argument(
        "--channel_size",
        type=int,
        default=128,
        help="Channel size (must be power of 2)",
    )
    parser.add_argument(
        "--num_simulations",
        type=int,
        default=200,
        help="Number of MCTS simulations per move",
    )
    parser.add_argument(
        "--batch_size",
        type=int,
        default=1000,
        help="Number of nodes to process at once",
    )
    parser.add_argument(
        "--eval_timeout_ms",
        type=int,
        default=1,
        help="Timeout for evaluation requests in milliseconds",
    )
    parser.add_argument(
        "--num_games",
        type=int,
        default=100,
        help="Number of games to play in evaluation",
    )
    parser.add_argument(
        "--model_path", type=str, default="model.pt", help="Path to the model file"
    )
    parser.add_argument(
        "--training_file",
        type=str,
        default="training_data.bin",
        help="File to store training data",
    )
    parser.add_argument(
        "--debug",
        action="store_true",
        help="Enable debug logging and intermediate state dumps",
    )
    parser.add_argument(
        "--temperature",
        type=float,
        default=1.0,
        help="Temperature parameter for move selection",
    )

    args = parser.parse_args()

    config = maz.get_config()
    config.channel_size = args.channel_size
    config.num_simulations = args.num_simulations
    config.batch_size = args.batch_size
    config.num_games = args.num_games
    config.model_path = args.model_path
    config.training_file = args.training_file
    config.debug = args.debug
    config.temperature = args.temperature

    print("Starting AlphaZero agent tournament...")

    maz.init_globals()
    maz.init_model()

    maz.start_evaluator_thread()

    agent1 = maz.RawModelAgent()
    agent2 = maz.RandomAgent()
    maz.run_agent_tournament(agent1, agent2)

    maz.stop_evaluator()


if __name__ == "__main__":
    main()
